### ISSUE TITLE
Fitting failed, unable to allocate memory

### ISSUE THREAD

**Issue by victor-felicitas** [biolab/orange3#5498]
I love Orange, but experience difficulties trying to run some analysis on a Kaggle Predict Sales dataset.
https://www.kaggle.com/c/competitive-data-science-predict-future-sales/data  

Whenever I try predicting daily sales with **linear regression** or **random forest** using factors as Price, ShopId and ItemID the Python freezes for ~30 minutes and shows error afterwards - "Fitting failed, unable to allocate XXX GiB for an array with shape YYY". 

Simple datasets like Titanic work fine. I also tried to limit number of rows in the Predict Sales training dataset to just a **few thousands** with Data Sample but somehow it doesn't help and any model just freezes.

I have installed Orange3-3.29.3-Miniconda-x86_64.exe (64 bit) and did not have any previous Python packages.
Have 16 gb of RAM. Amount of RAM used changes with the training dataset size, but it always takes ~30 minutes to unfreeze.

Any advice on what could have gone wrong would be appreciated!

**Comment by markotoplak** [comment]
Thanks for the report!

Could you share what XXX and what is YYY? :) Also, how many columns/rows does the original data set have? Are any of them nominal with a large number of valus? Could you perhaps also share the smallest dataset that is still problematic with us, preferable as an Orange .tab file, so that we can try to see what is happening?

For random forests: does it help if you limit the depth of the trees?

**Comment by victor-felicitas** [comment]
Thanks for the response!
Here is the smallest dataset I tested (only 10K rows).
Orange is able to process it in around an hour on my PC (random forest or linear regression).
Limiting tree depth to 4 doesn't help much.

[sales_train.zip](https://github.com/biolab/orange3/files/6698367/sales_train.zip)

**Comment by victor-felicitas** [comment]
Ok, it seems the issue is **a large number of different values for categorical features** compared to size of the dataset - i.e. in this Sales dataset there were thousands of different products in the dataset with ~2-10 sales records per each product, leading to the model choking.
Possible solution is grouping the products into larger groups.

I also **don't see any way to check for a number of individual values of  categorical features in Orange** - i.e., how many individual products we have now?.. For some reason feature statistics don't show it at all. Shouldn't it be demonstrated alongside max, min and median?

**Comment by markotoplak** [comment]
A large number of values for categorical features is expected to cause problems with scikit-learn based learners (which the two you used are). Scikit-learn only supports continuous features, therefore they are encoded to continuous with one-hot encoding, so the number of columns in that data set can really be large.

There is no escape from this for logistic regression, as it as a method assumes continuous inputs, but a proper implementation of random forests could handle categorical features directly, bypassing this problem. Unfortunately, Orange uses scikit-learn there which has its limitations. Precisely for that reason, Orange uses custom classification trees. Perhaps it could, someday, also switch to a different RF implementation.

**Comment by markotoplak** [comment]
@victor-felicitas, for a powerful ensemble-based technique in Orange try Gradient Boosting with `catboost`. That method does not need to expand categorical features. Be careful though, because other available boosting methods (scikit-learn and xgboost) will have the same limitations as random forests.

**Comment by victor-felicitas** [comment]
Indeed, catboost works, thanks @markotoplak .
